lib/src/layouts/yaru_master_detail_page.dart

```diff
@@ -81,43 +82,39 @@ class YaruMasterDetailPage extends StatefulWidget {
   /// An optional custom AppBar for the left pane.
   final PreferredSizeWidget? appBar;
 
-  /// An optional index of the initial page to show.
-  final int? initialIndex;
-
   /// Called when the user selects a page.
   final ValueChanged<int?>? onSelected;
 
   /// An optional controller that can be used to navigate to a specific index.
-  final ValueNotifier<int>? controller;
+  final YaruPageController? controller;
 
   @override
   _YaruMasterDetailPageState createState() => _YaruMasterDetailPageState();
 }
 
 class _YaruMasterDetailPageState extends State<YaruMasterDetailPage> {
-  var _index = -1;
-  var _previousIndex = 0;
-
   double? _previousPaneWidth;
+  late final YaruPageController _controller;
 
-  void _setIndex(int index) {
-    _previousIndex = _index;
-    _index = index;
-    widget.onSelected?.call(index == -1 ? null : index);
-  }
+  void _updateController() => _controller =
+      widget.controller ?? YaruPageController(length: widget.length);
 
   @override
   void initState() {
     super.initState();
-    _index = widget.initialIndex ?? -1;
+    _updateController();
+  }
+
+  @override
+  void dispose() {
+    if (widget.controller == null) _controller.dispose();
+    super.dispose();
   }
 
   @override
   void didUpdateWidget(covariant YaruMasterDetailPage oldWidget) {
     super.didUpdateWidget(oldWidget);
-    if (widget.initialIndex != oldWidget.initialIndex) {
-      _index = widget.initialIndex ?? -1;
-    }
+    if (widget.controller != oldWidget.controller) _updateController();
```

Comment:
missing `removeListener()` call on the old controller if not null

Reply:
`YaruMasterDetailPage` never adds any listener - that's done by either `YaruLandscapeLayout` or `YaruPortraitLayout`, which remove their listeners in `dispose()`. Is there anything that could go wrong?

Reply:
oh, i may have mixed up the layout names :)

Reply:
actually, are they both missing `didUpdateWidget()` here? can't the controller be different at init and dispose time if someone creates a controller inline and the whole thing gets rebuilt with a different controller?

Reply:
Yes, that can happen. Added `didUpdateWidget` to both - hope it works correctly now

Reply:
looks good i think